### ISSUE TITLE
Bash Pipeline Unit Test #1

### ISSUE THREAD

**Issue by rennergade** [Lind-Project/lind_project#140]
I've been testing with some of the example pipelines Brendan sent me, I'm currently using this one as a benchmark:

`grep '^.\{6\}$' words.txt | tr '[:upper:]' '[:lower:]' | sed 's/[abc]/2/g; s/[def]/3/g; s/[ghi]/4/g; s/[jkl]/5/g; s/[mno]/6/g; s/[pqrs]/7/g; s/[tuv]/8/g; s/[wxyz]/9/g' | sort | uniq | wc -l`

Creating this issue for debugging and reporting purposes.

**Comment by rennergade** [comment]
Running with a 1.2 MB wordlist (which I had from my AppSec class) Lind experiences a dramatic slowdown of 700ms to Native's 20ms.

This seems almost entirely due to `sort` spawning thousands of temp files to do the sorting, and it seems like Lind is compiling it with an option to use temp files for merge instead of memory. I'll need to dig into this.

This also makes the "coreutils programs should use big buffers" notion that I proposed at the meeting seem wrong. If you feed `grep` a large file, and it hits a high percentage of matches, it's output should theoretically be a large amount of small buffers. Same for sed etc...

**Comment by rennergade** [comment]
OTOH removing the `sort` from the pipeline has times of 88ms Lind vs 20 ms Native. This is completely dominated by program loading times at this point so pipes aren't really showing up at this small of a file size.

Trying the same pipeline with sort removed for a 1G file has Lind running at 5.6s and Native running at 2.2 seconds, so still a slowdown but much less pronounced.

Debugging futher by removing programs from the pipeline show that most of the slowdown is from the first pipe of grep to tr

**Comment by rennergade** [comment]
I did a bunch more testing here. I found that grep actually uses 2^16 buffers to read from the file, and then writes with small 2^13 buffers (it surprisingly does do buffering at all, which is good news!), while cat reads from file/writes to pipe both with 2^17. With our current implementation this tends to lend itself to why cat is faster than native but grep is slower.

OTOH I then tried grep (grep'ing all with .) vs cat on this 1GB wordlist file (as opposed to the 1GB of random bytes I tested before). Lind is actually slower than native here by about the same margin as grep above, but maintains the same 2^17 buffers.

This is all pretty weird, basically catting a 1GB file is faster in Lind when its random bytes, but native is ~3x faster when its a coherent wordlist. Lind pipes the 1GB at ~ 17s seconds in both examples, but native does 24s for the random bytes, and 5.5s for the wordlist.

Strace -fc of random bytes
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- --------------
57.31    2.335595      778531         3         1 wait4
 36.18    1.474631         184      8001           write
  6.48    0.264025           3     72051           read
  0.01    0.000485         161         3           execve
  0.00    0.000191           3        55        32 openat
  0.00    0.000151           3        47           mmap
```

strace -fc of wordlist

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- --------------
 64.24    0.561008      187002         3         1 wait4
 21.32    0.186153          22      8164           write
 14.06    0.122780           1     73511           read
  0.06    0.000560           6        83        61 stat
  0.06    0.000533         177         3           execve
  0.04    0.000353           7        47           mmap
```
So the writes are like 9x slower and reads are 3x slower for the random bytes. This seems eerily similar to the I/O weirdness we were seeing before but I still don't really have an explanation for why.

**Comment by rennergade** [comment]
After a fair bit of headscratching, I figured out that the overhead here for different layouts of 1G files is actually because of ho `wc` works and has nothing to do with the pipe! Doing a simple call to `wc` instead of `cat file | wc` gives the same overhead differences. I did not expect that

This means that the "cat to pipes is faster in Lind" results aren't really accurate, though my pipe between write/read programs still is. I believe there may be some slowdown due to reading from file since we have to touch the kernel multiple times there.

**Comment by rennergade** [comment]
Update: The slowness of the 1GB "random" file seems to because of locale issues. Replacing the file with random bytes base64 encoded, its shows the same behavior as the all A's example.

**Comment by rennergade** [comment]
I flamegraphed native vs lind for the `cat 1gbrandfile | wc` example. Native takes 2.7s while Lind takes 17 seconds, which is a larger slowdown than I see in other examples.
[Native](https://github.com/Lind-Project/lind_project/blob/graphs/catrand1GBnative.svg)
[Lind](https://github.com/Lind-Project/lind_project/blob/graphs/catrand1GBlind.svg

As another comparison here's `cat 1GBWordlist | grep '^.\{6\}$'` which finds 6 character words. The wordlist only has one six letter word so grep isn't doing much on the output side. This runs in 2.2 seconds native and 5.12 seconds in Lind, which is a more usual slowdown.

[Native](https://github.com/Lind-Project/lind_project/blob/graphs/catgrep1GBnative.svg)
[Lind](https://github.com/Lind-Project/lind_project/blob/graphs/catgrep1GBlind.svg)

Both of these seem to have the file read being the significant contributor to overhead, though the pipe speeds seem to be treading the line in terms of efficiency depending on the buffer size.

File reads are somewhat of an expensive process now. Not only do we use the repy readat mechanism which has its own seek lock (extra kernel calls) and does a few operations, but that reads into a Python string, which we still have to copy back into the user buffer.

**Comment by rennergade** [comment]
I did multiple tests where I just added extra cat's into the pipeline, for instance `cat 1gbrandfile | wc` became `cat 1gbrandfile | cat | cat | wc`

Native took 2.3 seconds with this modification (it was 2.2 with a single cat) and Lind took 17.2 seconds (it was ~17s with a single cat)

[Native](https://raw.githubusercontent.com/Lind-Project/lind_project/graphs/catcatcat1GBnative.svg)
[Lind](https://raw.githubusercontent.com/Lind-Project/lind_project/graphs/catcatcat1GBlind.svg)

This seems that adding additional pipelines, at least with cat-sized buffers, isn't a problem.

I ran `cat file` on both native and lind, redirecting stdout to a file (here I'm directing Lind's stdout to a native FS file, not redirecting bash's output within Lind to a lindfs file). Here, native takes 0.42 second when redirecting to a new file (important to note that this takes 2.3 seconds if the file already exists). Lind takes 0.86 seconds (regardless of if the file pre-exists or not).

In general I'm seeing this 2x slowdown with pipelines not using `wc`. `wc` uses smaller buffers for reads and I'm seeing an increased piperead overhead there, which would make sense.

**Comment by rennergade** [comment]
I spoke with Brendan about ways to try and improve Lind's file read/write performance before we try any nuclear option.

The first was to actually get the direct read/write from user buffers without having to do python intermediaries. I played around with this for a couple days and finally got it to work using ctypes, memorviews, and readinto, which was a good learning experience unto itself. Sadly, this didn't actually increase performance at all. It seems, just like with pipe write/reads, the python RPC is preventative. This is making reads about 2x as slow as native and writes about 1.5x as slow. Here's flamegraphs of just running the cat program to a file as a standalone in both native and lind for 1GB. Overall this is ~2x as slow which is what I'm seeing across the board.

[Native](https://raw.githubusercontent.com/Lind-Project/lind_project/graphs/catnat.svg)
[Lind](https://raw.githubusercontent.com/Lind-Project/lind_project/graphs/catlind.svg)

We also talked about trying to mmap the entire file and do reads and writes from memory to speed this up. I'm going to try this out next, but am somewhat skeptical still due to all the overhead being from RPC which this wouldn't alleviate.
